README.md
# androidlib

A suite of Android development-related functions.

## Legal

This project is open source under the [Apache Public License v2][1] and is developed by
[Axway, Inc](http://www.axway.com/) and the community. Please read the [`LICENSE`][1] file included
in this distribution for more information.

[1]: https://github.com/appcelerator/androidlib/blob/master/LICENSE


CHANGELOG.md
# v2.5.1 (Apr 15, 2021)

* fix: Switched from errno to code in connect catch due to breaking change in Node.js v13/14.

# v2.5.0 (Mar 29, 2019)

 * chore: Updated dependencies.

# v2.4.0 (Jan 29, 2019)

 * fix(connection): Fixed bug when executing adb command over a connection and `bufferUntilClose` is
   `true` and the results hang because of listening to `close` instead of `end` event.
 * chore: Upgraded to Gulp 4.
 * chore: Updated dependencies.

# v2.3.6 (Jun 14, 2018)

 * Fixed verbiage in invalid SDK when directory does not contain an emulator executable.

# v2.3.5 (Jun 12, 2018)

 * Added support for scanning the Android SDK `emulator` directory for the emulator executable.
   ([TIMOB-26126](https://jira.appcelerator.org/browse/TIMOB-26126))
 * Updated npm dependencies.
 * Removed `yarn.lock` from distribution.

# v2.3.4 (Apr 9, 2018)

 * Updated npm dependencies.

# v2.3.3 (Jan 10, 2018)

 * Fixed adb socket handling when adb goes away.

# v2.3.2 (Jan 4, 2018)

 * Updated npm dependencies, namely gawk@4.4.5.
 * Updated copyright year.

# v2.3.1 (Jan 4, 2018)

 * Prevent device changes from being emitted after stop tracking devices.
 * Clone internal devices array to strip gawk data before emitting.

# v2.3.0 (Dec 6, 2017)

 * Fixed bug where an ADB connection wouldn't wait for data for certain queries such as getting
   devices. This was especially noticeable when querying devices several times.

# v2.2.1 (Dec 6, 2017)

 * Added missing xmldom dependency.

# v2.2.0 (Dec 6, 2017)

 * Fixed regression where track devices would return emulators along with the connected devices.
 * Removed unnecessary device call to initialize track devices.
 * Added better error handling when getting a devices properties.
 * Added more debug logging.

# v2.1.1 (Dec 5, 2017)

 * Updated appcd npm deps to v1.0.0.

# v2.1.0 (Dec 4, 2017)

 * Fixed bug where multiple simultaneous calls to ADB-related functions would try to start multiple
   ADB instances at the same time.
 * Added `close` event to ADB connection and device tracking.
 * Fixed bug where updated list of devices would only be emitted if devices were connected or
   disconnected, but not if a device property was changed.
 * Track devices now emits the initial list of connected devices instead of future changes only.

# v2.0.11 (Nov 28, 2017)

 * Fixed bug in VirtualBox detection when trying to read non-existent `.xml` file.

# v2.0.10 (Nov 28, 2017)

 * Fixed incorrect Genymotion search paths option name.
 * Default `adb` executable in options to `null`.

# v2.0.9 (Nov 22, 2017)

 * Fixed Android NDK detection on Windows where certain NDK's `ndk-which` program did not include
   the `.cmd` extension.

# v2.0.8 (Nov 22, 2017)

 * Fixed detection of Android SDK add-ons that use a `manifest.ini` instead of a `source.properties`
   file.
 * Fixed `.ini` file line match regular expression to properly handle commented out lines.

# v2.0.7 (Nov 22, 2017)

 * Fixed populating emulator `basedOn`, `description`, `vendor`, and `version` properties.

# v2.0.6 (Nov 22, 2017)

 * Fixed bug with detecting connected devices as emulators.

# v2.0.5 (Nov 22, 2017)

 * Fixed bug with handling of Genymotion emulator ids.

# v2.0.4 (Nov 22, 2017)

 * Updated NPM dependencies.

# v2.0.3 (Nov 22, 2017)

 * Updated unnecessarily async emulator detection functions to sync.
 * Refactored Genymotion and VirtualBox detection.
 * Removed Genymotion detection dependency on `deployDir` and `vboxmanage`.

# v2.0.2 (Nov 20, 2017)

 * Fixed bug getting emulators when list of Android SDKs is empty.

# v2.0.1 (Nov 17, 2017)

 * Added missing macOS Genymotion mock files.
 * Fixed bug with custom Android SDK search paths.

# v2.0.0 (Nov 17, 2017)

 * Initial release of the v2 rewrite.
 * Updated code to ES2015.
 * Support for detecting SDKs, NDKs, devices, Android emulators, and Genymotion emulators.


lib/env.js
/**
 * environment related utils
 */
'use strict';
const fs = require('fs'),
	path = require('path');
let cached,
	cachedVersion;

exports.find = find;

// borrowed from node-appc since he doesn't export
function findSDK() {
	let dirs = process.platform === 'win32'
		? [ 'C:\\android-sdk', 'C:\\android', 'C:\\Program Files\\android-sdk', 'C:\\Program Files\\android', 'C:\\Program Files\\Android\\android-sdk',
			'C:\\Program Files (x86)\\android-sdk', 'C:\\Program Files (x86)\\android', 'C:\\Program Files (x86)\\Android\\android-sdk' ]
		: [ '/opt/android', '/opt/android-sdk', '/usr/android', '/usr/android-sdk', '/usr/local/android-sdk' ];
	const exe = process.platform === 'win32' ? 'android.exe' : 'android';

	for (let i = 0; i < dirs.length; i++) {
		if (fs.existsSync(dirs[i])) {
			return dirs[i];
		}
	}

	dirs = (process.env.PATH || '').split(process.platform === 'win32' ? ';' : ':');
	for (let i = 0; i < dirs.length; i++) {
		if (fs.existsSync(dirs[i].trim(), exe)) {
			return dirs[i];
		}
	}
}

function findNDK() {
	let dirs = process.platform === 'win32'
		? [ 'C:\\android-ndk', 'C:\\Program Files\\android-ndk', 'C:\\Program Files\\Android\\android-ndk',
			'C:\\Program Files (x86)\\android-ndk', 'C:\\Program Files (x86)\\Android\\android-ndk' ]
		: [ '/opt/android-ndk', '/usr/android-ndk', '/usr/local/android-ndk' ];
	const exe = process.platform === 'win32' ? 'ndk-build.cmd' : 'ndk-build';

	for (let i = 0; i < dirs.length; i++) {
		if (fs.existsSync(dirs[i])) {
			return dirs[i];
		}
	}

	dirs = (process.env.PATH || '').split(process.platform === 'win32' ? ';' : ':');
	for (let i = 0; i < dirs.length; i++) {
		if (fs.existsSync(dirs[i].trim(), exe)) {
			return dirs[i];
		}
	}
}

function getNDKToolchainDir() {
	const platform = (process.platform === 'win32' ? 'windows' : process.platform);
	let toolchain_dir = platform + '-x86';
	if (process.arch === 'x64') {
		toolchain_dir += '_64';
	}
	return toolchain_dir;
}

function getNDKToolchainArch(options) {
	if (options.arch === 'x86') {
		return 'x86';
	}
	if (options.arch === 'mips') {
		return 'mipsel-linux-android';
	}
	return 'arm-linux-androideabi';
}

function getShortArchName(options) {
	if (options.arch.match(/^arm/))	{
		return 'arm';
	}
	return options.arch;
}

function getLongArchName(options) {
	if (options.arch.match(/^arm/))	{
		if (options.arch === 'arm') {
			return 'armeabi';
		}
		return 'armeabi-v7a';
	}
	return options.arch;
}

function getNDKToolchain(options, ndk_dir, toolchain_arch, toolchain_dir, version) {
	return path.join(ndk_dir, 'toolchains', toolchain_arch + '-' + version, 'prebuilt', toolchain_dir);
}

function find(options, failIfNotFound) {
	if (cached) {
		options.sdk = options.sdk || cachedVersion;
		return cached;
	}
	let callback;
	if (typeof failIfNotFound === 'function') {
		callback = failIfNotFound;
		failIfNotFound = false;
	}
	options = options || {};
	const android_sdk = options['android-sdk'] || process.env.ANDROID_SDK_ROOT || process.env.ANDROID_SDK || findSDK(),
		android_ndk = options['android-ndk'] || process.env.ANDROID_NDK_ROOT || process.env.ANDROID_NDK || findNDK(),
		android_ndk_toolchain_ver = options['android-ndk-toolchain-version'] || process.env.ANDROID_NDK_TOOLCHAIN_VERSION || '4.8',
		android_ndk_toolchain_dir = getNDKToolchainDir(),
		android_ndk_toolchain_arch = getNDKToolchainArch(options),
		android_ndk_toolchain = getNDKToolchain(options, android_ndk, android_ndk_toolchain_arch, android_ndk_toolchain_dir, android_ndk_toolchain_ver),
		android_ndk_toolchain_llvm = path.join(android_ndk, 'toolchains', 'llvm', 'prebuilt', android_ndk_toolchain_dir),
		android_ndk_ar = path.join(android_ndk_toolchain, 'bin', android_ndk_toolchain_arch + '-ar'),
		android_ndk_clang = path.join(android_ndk_toolchain_llvm, 'bin', 'clang++'),
		android_ndk_platform_sdk = path.join(android_ndk, 'platforms', 'android-' + options.sdk, 'arch-' + getShortArchName(options)),
		android_ndk_libstdcpp = path.join(android_ndk, 'sources', 'cxx-stl', 'gnu-libstdc++', android_ndk_toolchain_ver),
		android_ndk_libstdcpplib = path.join(android_ndk_libstdcpp, 'libs', getLongArchName(options)),
		sdkversion = options.sdk,
		versions = sdkversion ? [ sdkversion ] : [ '19', '18', '17', '16', '15', '14' ],
		logger = options.logger;
	let found;

	if (!android_ndk && failIfNotFound) {
		logger.fatal('couldn\'t find Android NDK path. Please set environment to ANDROID_NDK');
	}

	if (android_sdk) {
		// try and find android
		for (let d = 0, dL = versions.length; d < dL; d++) {
			const ad = path.join(android_sdk, 'platforms', 'android-' + versions[d], 'android.jar');
			if (fs.existsSync(ad)) {
				found = ad;
				if (!sdkversion) {
					// if not specified explicitly, set it
					options.sdk = cachedVersion = versions[d];
				}
				break;
			}
		}

		if (found) {
			return (cached = {
				ndkPath: android_ndk,
				sdkPath: android_sdk,
				arch:  getShortArchName(options),
				archlib:  getLongArchName(options),
				toolchain: android_ndk_toolchain,
				toolchain_version: android_ndk_toolchain_ver,
				toolchain_dir: android_ndk_toolchain_dir,
				toolchain_llvm: android_ndk_toolchain_llvm,
				toolchain_arch: android_ndk_toolchain_arch,
				toolchain_ar: android_ndk_ar,
				toolchain_clang: android_ndk_clang,
				toolchain_platform_sdk: android_ndk_platform_sdk,
				toolchain_libstdcpp: android_ndk_libstdcpp,
				toolchain_libstdcpplib: android_ndk_libstdcpplib,
				jar: found
			});
		}
	}
	const msg = 'Android SDK not found. Please define environment ANDROID_SDK';
	if (failIfNotFound) {
		console.error(msg);
		process.exit(1);
	}
	if (callback) {
		return callback(msg);
	}
}
